test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java

```diff
@@ -0,0 +1,196 @@
+
+                System.err.print("\tasElement()");
+                checkEmptyAnnotations(typeUtils.asElement(returnType));
+
+                System.err.print("\tcapture()");
+                checkEmptyAnnotations(typeUtils.capture(returnType));
+
+                System.err.print("\terasure()");
+                checkEmptyAnnotations(typeUtils.erasure(returnType));
+
+                System.err.print("\tgetArrayType()");
+                ArrayType arrayType = typeUtils.getArrayType(returnType);
+                checkEmptyAnnotations(arrayType);
+                /*
+                 * "Annotations on the component type are preserved."
+                 */
+                checkEqualTypeAndAnnotations(returnType, arrayType.getComponentType());
+
+                if (!returnType.getKind().isPrimitive()) {
+                    /*
+                     * "Annotations on the bounds are preserved."
+                     */
+                    WildcardType wcType;
+                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(returnType, null));
+                    checkEqualTypeAndAnnotations(returnType, wcType.getExtendsBound());
+
+                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(null,       returnType));
+                    checkEqualTypeAndAnnotations(returnType, wcType.getSuperBound());
+                }
+
+                 System.out.println(returnType.getAnnotation(TestTypeAnnotation.class));
+                 System.out.println(returnType.getAnnotationsByType(TestTypeAnnotation.class).length);
+                 TestTypeAnnotation ta = requireNonNull(returnType.getAnnotation(TestTypeAnnotation.class),
+                                                        returnType.toString());
+
+                System.err.println();
+                System.err.println();
+            }
+
+            if (failures > 0)
+                throw new RuntimeException(failures + " failures occured.");
+        }
+        return true;
+    }
+
+    void testVacuous(TypeMirror tm ) {
+        try {
+            var result = vacuousTypes.stripAnnotations(tm);
+            messager.printError("Unexpected non-exceptional result returned" +  result);
+        } catch(UnsupportedOperationException uoe) {
+            ; // Expected
+        }
+    }
+
+    private int failures = 0;
+
+    void checkExpectedTypeAnnotations(AnnotatedConstruct ac, TypeMirror expectedAnnotation) {
+        List<? extends AnnotationMirror> annotations = ac.getAnnotationMirrors();
+        if (annotations.size() != 1) {
+            failures++;
+            System.err.println("\t\t\tUnexpected annotations size: " + annotations.size());
+        } else if (!typeUtils.isSameType(annotations.get(0).getAnnotationType(), expectedAnnotation)) {
+            failures++;
+            System.err.println("\t\t\tUnexpected annotations type: " + annotations);
+        }
+    }
+
+    void checkEmptyAnnotations(AnnotatedConstruct ac) {
```

Comment:
I think it would be valuable to expand the assertion to visit contained types, and assert that none of the contained types have annotations:

```diff
diff --git a/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java b/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
index c627f41ba0b..8bb3c7cd3a8 100644
--- a/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
+++ b/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
@@ -82,10 +82,10 @@ public boolean process(Set<? extends TypeElement> annotations,
 
                 System.err.print("\tgetArrayType()");
                 ArrayType arrayType = typeUtils.getArrayType(returnType);
-                checkEmptyAnnotations(arrayType);
                 /*
                  * "Annotations on the component type are preserved."
                  */
+                checkEmptyAnnotations((AnnotatedConstruct) arrayType);
                 checkEqualTypeAndAnnotations(returnType, arrayType.getComponentType());
 
                 if (!returnType.getKind().isPrimitive()) {
@@ -93,10 +93,14 @@ public boolean process(Set<? extends TypeElement> annotations,
                      * "Annotations on the bounds are preserved."
                      */
                     WildcardType wcType;
-                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(returnType, null));
+                    System.err.print("\tgetWildcardType(returnType, null)");
+                    wcType = typeUtils.getWildcardType(returnType, null);
+                    checkEmptyAnnotations((AnnotatedConstruct) wcType);
                     checkEqualTypeAndAnnotations(returnType, wcType.getExtendsBound());
 
-                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(null,       returnType));
+                    System.err.print("\tgetWildcardType(null, returnType)");
+                    wcType = typeUtils.getWildcardType(null, returnType);
+                    checkEmptyAnnotations((AnnotatedConstruct) wcType);
                     checkEqualTypeAndAnnotations(returnType, wcType.getSuperBound());
                 }
 
@@ -158,6 +162,54 @@ void checkEmptyAnnotations(AnnotatedConstruct ac) {
         }
     }
 
+    void checkEmptyAnnotations(TypeMirror ac) {
+        System.err.println("\t" + ac);
+        if (ac == null) {
+            return;
+        }
+        new SimpleTypeVisitor14<Void, Void>() {
+            @Override
+            protected Void defaultAction(TypeMirror t, Void o) {
+                checkEmptyAnnotations((AnnotatedConstruct) t);
+                return null;
+            }
+
+            @Override
+            public Void visitArray(ArrayType t, Void o) {
+                scan(t.getComponentType());
+                return super.visitArray(t, o);
+            }
+
+            @Override
+            public Void visitDeclared(DeclaredType t, Void o) {
+                scan(t.getEnclosingType());
+                t.getTypeArguments().stream().forEach(this::scan);
+                return super.visitDeclared(t, o);
+            }
+
+            @Override
+            public Void visitTypeVariable(TypeVariable t, Void o) {
+                // the bounds correspond to the type variable declaration, not its use
+                // scan(t.getUpperBound());
+                // scan(t.getLowerBound());
+                return super.visitTypeVariable(t, o);
+            }
+
+            @Override
+            public Void visitWildcard(WildcardType t, Void o) {
+                scan(t.getExtendsBound());
+                scan(t.getSuperBound());
+                return super.visitWildcard(t, o);
+            }
+
+            private void scan(TypeMirror t) {
+                if (t != null) {
+                    visit(t);
+                }
+            }
+        }.visit(ac);
+    }
+
     void checkEqualTypeAndAnnotations(TypeMirror tm1, TypeMirror tm2) {
         if (!typeUtils.isSameType(tm1, tm2)) {
             failures++;

```

Reply:
> I think it would be valuable to expand the assertion to visit contained types, and assert that none of the contained types have annotations:
> 
> ```diff
> diff --git a/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java b/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
> index c627f41ba0b..8bb3c7cd3a8 100644
> --- a/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
> +++ b/test/langtools/tools/javac/processing/model/util/types/TestAnnotationStripping.java
> @@ -82,10 +82,10 @@ public boolean process(Set<? extends TypeElement> annotations,
>  
>                  System.err.print("\tgetArrayType()");
>                  ArrayType arrayType = typeUtils.getArrayType(returnType);
> -                checkEmptyAnnotations(arrayType);
>                  /*
>                   * "Annotations on the component type are preserved."
>                   */
> +                checkEmptyAnnotations((AnnotatedConstruct) arrayType);
>                  checkEqualTypeAndAnnotations(returnType, arrayType.getComponentType());
>  
>                  if (!returnType.getKind().isPrimitive()) {
> @@ -93,10 +93,14 @@ public boolean process(Set<? extends TypeElement> annotations,
>                       * "Annotations on the bounds are preserved."
>                       */
>                      WildcardType wcType;
> -                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(returnType, null));
> +                    System.err.print("\tgetWildcardType(returnType, null)");
> +                    wcType = typeUtils.getWildcardType(returnType, null);
> +                    checkEmptyAnnotations((AnnotatedConstruct) wcType);
>                      checkEqualTypeAndAnnotations(returnType, wcType.getExtendsBound());
>  
> -                    checkEmptyAnnotations(wcType = typeUtils.getWildcardType(null,       returnType));
> +                    System.err.print("\tgetWildcardType(null, returnType)");
> +                    wcType = typeUtils.getWildcardType(null, returnType);
> +                    checkEmptyAnnotations((AnnotatedConstruct) wcType);
>                      checkEqualTypeAndAnnotations(returnType, wcType.getSuperBound());
>                  }
>  
> @@ -158,6 +162,54 @@ void checkEmptyAnnotations(AnnotatedConstruct ac) {
>          }
>      }
>  
> +    void checkEmptyAnnotations(TypeMirror ac) {
> +        System.err.println("\t" + ac);
> +        if (ac == null) {
> +            return;
> +        }
> +        new SimpleTypeVisitor14<Void, Void>() {
> +            @Override
> +            protected Void defaultAction(TypeMirror t, Void o) {
> +                checkEmptyAnnotations((AnnotatedConstruct) t);
> +                return null;
> +            }
> +
> +            @Override
> +            public Void visitArray(ArrayType t, Void o) {
> +                scan(t.getComponentType());
> +                return super.visitArray(t, o);
> +            }
> +
> +            @Override
> +            public Void visitDeclared(DeclaredType t, Void o) {
> +                scan(t.getEnclosingType());
> +                t.getTypeArguments().stream().forEach(this::scan);
> +                return super.visitDeclared(t, o);
> +            }
> +
> +            @Override
> +            public Void visitTypeVariable(TypeVariable t, Void o) {
> +                // the bounds correspond to the type variable declaration, not its use
> +                // scan(t.getUpperBound());
> +                // scan(t.getLowerBound());
> +                return super.visitTypeVariable(t, o);
> +            }
> +
> +            @Override
> +            public Void visitWildcard(WildcardType t, Void o) {
> +                scan(t.getExtendsBound());
> +                scan(t.getSuperBound());
> +                return super.visitWildcard(t, o);
> +            }
> +
> +            private void scan(TypeMirror t) {
> +                if (t != null) {
> +                    visit(t);
> +                }
> +            }
> +        }.visit(ac);
> +    }
> +
>      void checkEqualTypeAndAnnotations(TypeMirror tm1, TypeMirror tm2) {
>          if (!typeUtils.isSameType(tm1, tm2)) {
>              failures++;
> ```

I've added a variation of the new code to the PR. The checks I added are in line with the properties discussed by the API changes in Types. Thanks.